exporter/metrics.go

```diff
@@ -216,24 +237,39 @@ func makeMetrics(prefix string, m bson.M, labels map[string]string) []prometheus
 	for k, val := range m {
 		switch v := val.(type) {
 		case bson.M:
-			res = append(res, makeMetrics(prefix+k, v, labels)...)
+			res = append(res, makeMetrics(prefix+k, v, labels, compatibleMode)...)
 		case map[string]interface{}:
-			res = append(res, makeMetrics(prefix+k, v, labels)...)
+			res = append(res, makeMetrics(prefix+k, v, labels, compatibleMode)...)
 		case primitive.A:
 			v = []interface{}(v)
-			res = append(res, processSlice(prefix, k, v)...)
+			res = append(res, processSlice(prefix, k, v, compatibleMode)...)
 		case []interface{}:
 			continue
 		default:
-			metric, err := makeRawMetric(prefix, k, v, labels)
+			rm, err := makeRawMetric(prefix, k, v, labels)
 			if err != nil {
-				metric = prometheus.NewInvalidMetric(prometheus.NewInvalidDesc(err), err)
+				invalidMetric := prometheus.NewInvalidMetric(prometheus.NewInvalidDesc(err), err)
+				res = append(res, invalidMetric)
+				continue
 			}
 
 			// makeRawMetric returns a nil metric for some data types like strings
 			// because we cannot extract data from all types
-			if metric != nil {
-				res = append(res, metric)
+			if rm == nil {
+				continue
```

Comment:
Just a question: shouldn't we log them as invalid metrics as well?

Reply:
No. The commands response, for example, getDiagnosticData response, has metrics and some other information like strings with the database version. Those string (and some other types) are not error, they are just meaningless values.